tests.rb

```diff
@@ -0,0 +1,149 @@
+    end
+
+    logdna_thread.join
+    server_thread.join
+  end
+
+  def info_method(port)
+    options = get_options(port)
+    logdna_thread = Thread.start do
+      logger = Logdna::Ruby.new("pp", options)
+      logger.info(@@log_line)
+    end
+
+    server_thread = Thread.start do
+      sor = TestServer.new
+      recieved_data = sor.start_server(port)
+
+      assert_equal(recieved_data[:ls][0][:line], @@log_line)
+      assert_equal(recieved_data[:ls][0][:app], options[:app])
+      assert_equal(recieved_data[:ls][0][:level], "INFO")
+      assert_equal(recieved_data[:ls][0][:env], options[:env])
+    end
+
+    logdna_thread.join
+    server_thread.join
+  end
+
+  def debug_method(port)
+    options = get_options(port)
+    logdna_thread = Thread.start do
+      logger = Logdna::Ruby.new("pp", options)
+      logger.debug(@@log_line)
+    end
+
+    server_thread = Thread.start do
+      sor = TestServer.new
+      recieved_data = sor.start_server(port)
+
+      assert_equal(recieved_data[:ls][0][:line], @@log_line)
+      assert_equal(recieved_data[:ls][0][:app], options[:app])
+      assert_equal(recieved_data[:ls][0][:level], "DEBUG")
+      assert_equal(recieved_data[:ls][0][:env], options[:env])
+    end
+
+    logdna_thread.join
+    server_thread.join
+  end
+
+  def fatal_method(port)
+    options = get_options(port)
+    logdna_thread = Thread.start do
+      logger = Logdna::Ruby.new("pp", options)
+      logger.fatal(@@log_line)
+    end
+
+    server_thread = Thread.start do
+      sor = TestServer.new
+      recieved_data = sor.start_server(port)
+
+      assert_equal(recieved_data[:ls][0][:line], @@log_line)
+      assert_equal(recieved_data[:ls][0][:app], options[:app])
+      assert_equal(recieved_data[:ls][0][:level], "FATAL")
+      assert_equal(recieved_data[:ls][0][:env], options[:env])
+    end
+
+    logdna_thread.join
+    server_thread.join
+  end
+
+  # Should retry to connect and preserve the failed line
+  def fatal_method_not_found(port)
+    second_line = " second line"
+    options = get_options(port)
+    logdna_thread = Thread.start do
+      logger = Logdna::Ruby.new("pp", options)
+      logger.fatal(@@log_line)
+      logger.fatal(second_line)
+    end
+
+    server_thread = Thread.start do
+      sor = TestServer.new
+      recieved_data = sor.return_not_found_res(port)
+      # The order of recieved lines is unpredictable.
+      recieved_lines = [
+        recieved_data[:ls][0][:line],
+        recieved_data[:ls][1][:line]
+      ]
+
+      assert_includes(recieved_lines, @@log_line)
+      assert_includes(recieved_lines, second_line)
+
+      assert_equal(recieved_data[:ls][0][:app], options[:app])
+      assert_equal(recieved_data[:ls][0][:level], "FATAL")
+      assert_equal(recieved_data[:ls][0][:env], options[:env])
+    end
+
+    logdna_thread.join
+    server_thread.join
+  end
+
+  def test_all
+    warn_method(2000)
+    info_method(2001)
+    fatal_method(2002)
```

Comment:
Along the lines of my comment above, these would be:
```
  test_log_level('warn', 2000, [expected_output])
  test_log_level('info', 2001, [expected_output])
  test_log_level('fatal', 2002, [expected_output])
```
etc

Reply:
Danny, I think converting a string to a method would be too complicated for a testing script. I prefer to keep testing code simple.

Reply:
I agree with keeping testing code simple, but the repeated copied code is a strong code smell that something is wrong. The way I suggest feels most Ruby-like to me, but if you don't like it, figure out a way to pull out the common code into a shared function.